databricks_cli/pipelines/api.py

```diff
@@ -0,0 +1,187 @@
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from hashlib import sha1
+import os
+
+from six.moves import urllib
+
+from databricks_cli.sdk import DeltaPipelinesService
+from databricks_cli.dbfs.api import DbfsApi
+from databricks_cli.dbfs.dbfs_path import DbfsPath
+
+
+# These imports are specific to the credentials part
+from databricks_cli.configure.config import get_profile_from_context
+from databricks_cli.configure.provider import get_config, ProfileConfigProvider
+from databricks_cli.utils import InvalidConfigurationError
+
+BUFFER_SIZE = 1024 * 64
+base_pipelines_dir = 'dbfs:/pipelines/code/'
+
+
+class PipelinesApi(object):
+    def __init__(self, api_client):
+        self.client = DeltaPipelinesService(api_client)
+        self.dbfs_client = DbfsApi(api_client)
+
+    def deploy(self, spec, headers=None):
+        lib_objects = LibraryObject.convert_from_libraries(spec.get('libraries', []))
+        local_lib_objects, rest_lib_objects = \
+            self._partition_libraries_and_extract_local_paths(lib_objects)
+        remote_lib_objects = list(map(lambda llo:
+                                      LibraryObject(llo.lib_type, self._get_hashed_path(llo.path)),
+                                      local_lib_objects))
+        upload_files = self._get_files_to_upload(local_lib_objects, remote_lib_objects)
+
+        for llo, rlo in upload_files:
+            try:
+                self.dbfs_client.put_file(llo.path, rlo.path, False)
+            except Exception as e:
+                raise RuntimeError('Error \'{}\' while uploading {}'.format(e, llo.path))
+
+        spec['libraries'] = LibraryObject.convert_to_libraries(rest_lib_objects +
+                                                               remote_lib_objects)
+        spec['credentials'] = self._get_credentials_for_request()
+        self.client.client.perform_query('PUT',
+                                         '/pipelines/{}'.format(spec['id']),
+                                         data=spec,
+                                         headers=headers)
+
+    def delete(self, pipeline_id, headers=None):
+        self.client.delete(pipeline_id, self._get_credentials_for_request(), headers)
+
+    @staticmethod
+    def _partition_libraries_and_extract_local_paths(lib_objects):
+        """
+        Partitions the given set of libraries into local and remote by checking uri scheme
```

Comment:
Can you describe the criteria of what makes a local/remote library?

Reply:
Local libraries are present on the user machine which can be either prefixed by file: or no scheme at all (described in the doc)
All other libraries in the spec are remote. Currently, they should be dbfs paths, but this will probably be expanded in the future. I'll update the documentation to briefly describe this.

Reply:
Ah, I see why I'm confused. I think you overload the term "remote" in your PR here. One describes the remote versions of the local libraries you are trying to upload and the other describes libraries which actually sit on S3/DBFS. Can we make sure that we don't overload the word `remote` in docstrings and variable names?

Reply:
It might be more clear if you split PipelinesApi.deploy into three clear steps
1. Identify what are the local libs
2. Write a function called `uploadLocalLibs` which identifies where the local libs should go and which ones should be uploaded. The function can return the uploaded versions of the local libs.
3. Replace the libraries field with the union of return of 2 and remotes identified in 1.

Reply:
Changed the deploy flow to what you mentioned. 
To decrease confusion further, I've renamed "rest" to "external".